modules/mastodon/index.js

```diff
@@ -0,0 +1,288 @@
+ * adding them to the "blocked" key in this module's config.json.
+ */
+
+var Cacheman = require('cacheman')
+  , contentType = require('content-type')
+  , execSync = require('child_process').execSync
+  , htmlToTextMod = require('html-to-text')
+  , robotsParser = require('robots-parser')
+  , robotsTxtCache = new Cacheman()
+  , sprintf = require('extsprintf').sprintf
+  , util = require('util')
+  , urlMod = require('url');
+
+/* variables defined in module.exports.init */
+var bot = null
+  , log = null
+  , moduleConfig = null
+  , request = null;
+
+function robotRequest(options, fn) {
+  var url = options.url;
+  var userAgent = moduleConfig != null && moduleConfig.userAgent || 'EuIrcBot';
+  var robotsTxtUrl = urlMod.resolve(url, '/robots.txt');
+
+  var handle = function (robotsTxt) {
+    robots = robotsParser(robotsTxtUrl, robotsTxt);
+    /* `User-agent: *` rules are generally overzealous. We're not a
+     * search engine spider. */
+    delete robots._rules['*'];
+
+    if (robots.isAllowed(url, userAgent)) {
+      request(options, fn);
+    } else {
+      fn(new Error(util.format('URL access disallowed by %s: %s', robotsTxtUrl, url)));
+    }
+  };
+
+  robotsTxtCache.get(robotsTxtUrl, function (err, robotsTxt) {
+    if (err) {
+      fn(new Error('robots.txt cache error: ' + err.message));
+    } else {
+      if (!robotsTxt) {
+        request(robotsTxtUrl, function (err, response, robotsTxt) {
+          if (err) {
+            fn(new Error(util.format('failed to fetch %s: %s', robotsTxtUrl, err.message)));
+          } else {
+            /* Determine TTL for this robots.txt. Google says that requests are
+             * "generally cached for up to one day" but that it may adjust that
+             * based on max-age Cache-Control headers, which seems reasonable. */
+            var ttl = 86400;
+            if (response.headers['cache-control']) {
+              var match = response.headers['cache-control'].match(/\bmax-age=([0-9]+)\b/);
+              if (match) {
+                /* Clamp this value to between (300, 604800) to guard against
+                 * unreasonable configuration / bugs in this code. */
+                ttl = Math.min(Math.max(parseInt(match[1], 10), 300), 604800);
+              }
+            }
+            robotsTxtCache.set(robotsTxtUrl, robotsTxt, ttl, function (err) {
+              if (err) {
+                fn(new Error('robots.txt cache error: ' + err.message));
+              } else {
+                handle(robotsTxt);
+              }
+            });
+          }
+        });
+      } else {
+        handle(robotsTxt);
+      }
+    }
+  });
+}
+
+function defaultConfig() {
+  return {'instances': require('./default_instances.json')};
+}
+
+function writeConfig() {
+  bot.writeDataFile('config.json', JSON.stringify(moduleConfig), function(err) {
+    if (err)
+      log.error('error writing mastodon module data: ' + err);
+  });
+}
+
+function checkHostname(hostname, fn) {
```

Comment:
This doesn't do anything async, so it can just return a bool rather than deal with callbacks.

Reply:
I think initially I had checkHostname verifying the validity of a Mastodon instance, but that got moved to forceCheckHostname, so this makes sense.